src/Microsoft.ML.Core/Data/VBuffer.cs

```diff
@@ -458,5 +458,12 @@ public T GetItemOrDefault(int slot)
                 return Values[index];
             return default(T);
         }
+
+        public override string ToString()
+        {
+            if (IsDense)
+                return $"Dense vector of size {Length}";
+            return $"Sparse vector of size {Length}, {Count} explicit values";
+        }
```

Comment:
I do not like this `ToString` method. *Generally* two string are not merely vaguely descriptive strings, they often represent the object in some sense. This type of descriptive rather than representative text would be more properly expressed in a `DebuggerDisplayAttribute`.

Reply:
Unfortunately, it didn't work well for me. You will only get the debugger display if you are looking at one VBuffer in isolation. A `KeyValuePair<string, VBuffer>` for example, used to be rendered as `{"foo", {Sparse vector of size 100, 10 explicit values}}`, and if I switch to `DebuggerDisplay` it's becoming `{"foo", "VBuffer['1{ReadOnlyMemory['1char]]}`

---
In reply to: [230566619](https://github.com/dotnet/machinelearning/pull/1518#discussion_r230566619) [](ancestors = 230566619)

Reply:
@mikem8361 - Do you know who on the VS Debugger team to reach out to for this issue?  Is the above behavior expected?

Reply:
KeyValuePair doesn't have a DebuggerDisplay, so what you are asking for wouldn't work. But (and maybe this is why KeyValuePair doesn't have a debugger display), it looks like the EE will not recursively apply DebuggerDisplay even if it did. I am not quite sure why, but you are welcome to open a bug or feedback ticket.